Add an article how to implement a log handler (#364)
apple/swift-log: Sources/Logging/Docs.docc/ImplementingALogHandler.md

# Implementing a log handler

> A guide for implementing custom logging backends that conform to the
> ``LogHandler`` protocol.

Reviewer: ```suggestion
Create a custom logging backend that provides logging services for your apps and libraries.
```

You generally don't want to use code voice or links to symbols in abstracts within DocC, as they don't reflect the links correctly, since they're rendered in multiple areas. It's better to spell it out narratively for that first sentence aligned with the title.

I also rephrased the abstract in the suggestion to describe what this article covers (how to conform, etc) and why you'd do it, not mentioning the specific protocol in the abstract.

## Overview

> To become a compatible logging backend that all `SwiftLog` consumers can use,
> you need to fulfill a few requirements.

Reviewer: ```suggestion
To become a compatible logging backend that any `SwiftLog` consumer can use,
you need to fulfill a few requirements, primarily conforming to the ``LogHandler`` protocol.
```

### Implementation Requirements

#### Value Type Semantics

Your log handler **must be a `struct`** and treat its configuration as value
types. This ensures that changes to one logger don't affect others.

> To ensure your handler has proper value semantics ensure that it passes this
> test:

Reviewer: ```suggestion
To verify that your handler reflects value semantics ensure that it passes this
test:
```

reworded to not use "ensure" twice in the sentence

```swift
@Test
func logHandlerValueSemantics() {
    LoggingSystem.bootstrap(MyLogHandler.init)
    var logger1 = Logger(label: "first logger")
    logger1.logLevel = .debug
    logger1[metadataKey: "only-on"] = "first"
    
    var logger2 = logger1
    logger2.logLevel = .error                  // Must not affect logger1
    logger2[metadataKey: "only-on"] = "second" // Must not affect logger1
    
    // These expectations must pass
    #expect(logger1.logLevel == .debug)
    #expect(logger2.logLevel == .error)
    #expect(logger1[metadataKey: "only-on"] == "first")
    #expect(logger2[metadataKey: "only-on"] == "second")
}

> ```

Reviewer 2: that's a nice idea, to show that 👍

> >Note: In special cases, it is acceptable that a log handler provides

Reviewer 2: ```suggestion
> Note: In special cases, it is acceptable that a log handler provides
```

> some global log level override that may affect all log handlers created.

Reviewer 2: ```suggestion
> some global log level override that may affect all log handlers created.
```

> >Note: In special cases, it is acceptable that a log handler provides
> some global log level override that may affect all log handlers created.

Reviewer: ```suggestion
> Note: In special cases, it is acceptable for a log handler to provide
> global log level overrides that may affect all log handlers created.
```

Needs a space in the callout

        function: String,
        line: UInt
    ) {
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let levelString = level.rawValue.uppercased()
        
        // Merge handler metadata with message metadata
        let combinedMetadata = self.metadata
            .merging(metadata ?? [:]) { _, new in new }
        
        // Format metadata
        let metadataString = combinedMetadata.map { "\($0.key)=\($0.value)" }.joined(separator: ",")
        
        // Create log line and print to console
        let logLine = "\(label) \(timestamp) \(levelString) [\(metadataString)]: \(message)"
        print(logLine)
    }
    
    public subscript(metadataKey key: String) -> Logger.Metadata.Value? {
        get {
            return self.metadata[key]
        }
        set {
            self.metadata[key] = newValue
        }
    }
}

```

### Advanced Features

#### Metadata Providers

Metadata providers allow you to dynamically add contextual information to all
log messages without explicitly passing it each time. Common use cases include
request IDs, user sessions, or trace contexts that should be included in logs
throughout a request's lifecycle.

> Support dynamic metadata injection:

Reviewer 2: Not sure if necessary
```suggestion
```

```swift
import Foundation
import Logging

public struct PrintLogHandler: LogHandler {
    private let label: String
    public var logLevel: Logger.Level = .info
    public var metadata: Logger.Metadata = [:]
    public var metadataProvider: Logger.MetadataProvider?
    
    public init(label: String) {
        self.label = label
    }
    
    public func log(
        level: Logger.Level,
        message: Logger.Message,
        metadata: Logger.Metadata?,
        source: String,
        file: String,
        function: String,
        line: UInt
    ) {
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let levelString = level.rawValue.uppercased()
        
        // Get provider metadata
        let providerMetadata = metadataProvider?.get() ?? [:]
        
        // Merge all metadata sources
        let combinedMetadata = providerMetadata
            .merging(self.metadata) { _, new in new }

>             .merging(metadata ?? [:]) { _, new in new }

Reviewer 2: heh I wonder if this should show a more efficient way if we're going to show people a copy-paste-ready snippet?

Our own one does:  https://github.com/apple/swift-log/blob/main/Sources/Logging/Logging.swift#L1404

        
        // Format metadata
        let metadataString = combinedMetadata.map { "\($0.key)=\($0.value)" }.joined(separator: ",")
        
        // Create log line and print to console
        let logLine = "\(label) \(timestamp) \(levelString) [\(metadataString)]: \(message)"
        print(logLine)
    }
    
    public subscript(metadataKey key: String) -> Logger.Metadata.Value? {
        get {
            return self.metadata[key]
        }
        set {
            self.metadata[key] = newValue
        }
    }
}
```

## Performance Considerations

> 1. **Avoid blocking**: Don't block the calling thread for I/O operations
> 2. **Lazy evaluation**: Remember that messages and metadata are autoclosures
> 3. **Memory efficiency**: Don't hold onto large amounts of messages

Reviewer: ```suggestion
1. **Avoid blocking**: Don't block the calling thread for I/O operations.
2. **Lazy evaluation**: Remember that messages and metadata are autoclosures.
3. **Memory efficiency**: Don't hold onto large amounts of messages.
```

Make each of these complete sentences

# Implementing a log handler

Create a custom logging backend that provides logging services for your apps
and libraries.

## Overview

To become a compatible logging backend that any `SwiftLog` consumer can use,
you need to fulfill a few requirements, primarily conforming to the
``LogHandler`` protocol.

> ### Implementation Requirements
> 
> #### Value Type Semantics

Reviewer: ```suggestion
### Implement with Value Type Semantics
```

This is a redundant header right underneath the section header - I'd recommend combining them.

Reviewer 3: Also should be sentence cased all the way

Your log handler **must be a `struct`** and exhibit value semantics. This
ensures that changes to one logger don't affect others.

To verify that your handler reflects value semantics ensure that it passes this
test:

```swift
@Test
func logHandlerValueSemantics() {
    LoggingSystem.bootstrap(MyLogHandler.init)
    var logger1 = Logger(label: "first logger")
    logger1.logLevel = .debug
    logger1[metadataKey: "only-on"] = "first"
    
    var logger2 = logger1
    logger2.logLevel = .error                  // Must not affect logger1
    logger2[metadataKey: "only-on"] = "second" // Must not affect logger1
    
    // These expectations must pass
    #expect(logger1.logLevel == .debug)
    #expect(logger2.logLevel == .error)
    #expect(logger1[metadataKey: "only-on"] == "first")
    #expect(logger2[metadataKey: "only-on"] == "second")
}
```

> Note: In special cases, it is acceptable for a log handler to provide
> global log level overrides that may affect all log handlers created.

> ## Example Implementation

Reviewer: ```suggestion
### Example Implementation
```

I think this makes more sense as a level 3 header, rather than jumping "back up" to level 2 - that puts it on the same par as the other level 3 headers, which from a ToC point of view seems better

Reviewer 3: Also sentence case

        let levelString = level.rawValue.uppercased()
        
        // Merge handler metadata with message metadata
        let combinedMetadata = Self.prepareMetadata(
            base: self.metadata
            explicit: metadata
        )
        
        // Format metadata
        let metadataString = combinedMetadata.map { "\($0.key)=\($0.value)" }.joined(separator: ",")
        
        // Create log line and print to console
        let logLine = "\(label) \(timestamp) \(levelString) [\(metadataString)]: \(message)"
        print(logLine)
    }
    
    public subscript(metadataKey key: String) -> Logger.Metadata.Value? {
        get {
            return self.metadata[key]
        }
        set {
            self.metadata[key] = newValue
        }
    }

    static func prepareMetadata(
        base: Logger.Metadata,
        explicit: Logger.Metadata?
    ) -> Logger.Metadata? {
        var metadata = base

        guard let explicit else {
            // all per-log-statement values are empty
            return metadata
        }

        metadata.merge(explicit, uniquingKeysWith: { _, explicit in explicit })

        return metadata
    }
}

```

> ### Advanced Features

Reviewer 3: ```suggestion
### Advanced features
```

> #### Metadata Providers

Reviewer 3: ```suggestion
#### Metadata providers
```

        
        // Create log line and print to console
        let logLine = "\(label) \(timestamp) \(levelString) [\(metadataString)]: \(message)"
        print(logLine)
    }
    
    public subscript(metadataKey key: String) -> Logger.Metadata.Value? {
        get {
            return self.metadata[key]
        }
        set {
            self.metadata[key] = newValue
        }
    }

    static func prepareMetadata(
        base: Logger.Metadata,
        provider: Logger.MetadataProvider?,
        explicit: Logger.Metadata?
    ) -> Logger.Metadata? {
        var metadata = base

        let provided = provider?.get() ?? [:]

        guard !provided.isEmpty || !((explicit ?? [:]).isEmpty) else {
            // all per-log-statement values are empty
            return metadata
        }

        if !provided.isEmpty {
            metadata.merge(provided, uniquingKeysWith: { _, provided in provided })
        }

        if let explicit = explicit, !explicit.isEmpty {
            metadata.merge(explicit, uniquingKeysWith: { _, explicit in explicit })
        }

        return metadata
    }
}
```

> ## Performance Considerations

Reviewer: ```suggestion
### Performance Considerations
```

Reviewer 3: Also sentence case
